File: Data Structures & Algorithms/maximum-subarray/submission-4.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxSubArray(nums) {
        let n =nums.length ;
        const dfs = (i,flag)=>{
            if(i==n)
              return flag ? 0:-10000;
            return flag ? Math.max(dfs(i+1,true)+nums[i],0) : Math.max(dfs(i+1,false),nums[i]+dfs(i+1,true)) ;
        }
        return dfs(0,false);
    }
}

File: Data Structures & Algorithms/house-robber/submission-4.cpp
class Solution {
public:
    int dfs(int i , vector<int>& nums,vector<int>& memo){
        if(i>=nums.size())
              return 0 ;
        if(memo[i]!=-1)
              return memo[i];
        memo[i]=max(dfs(i+1,nums,memo),nums[i]+dfs(i+2,nums,memo));
        return memo[i];
    }
    int rob(vector<int>& nums) {
        vector<int> memo(nums.size(),-1);
        return dfs(0,nums,memo);
    }
};
